core/io/zip_io.cpp

```diff
@@ -30,6 +30,58 @@
 
 #include "zip_io.h"
 
+#include "core/templates/local_vector.h"
+
+int godot_unzip_get_current_file_info(unzFile p_zip_file, unz_file_info64 &r_file_info, String &r_filepath) {
+	auto try_get_current_file_info = [&](char *p_szFilePathBuffer = nullptr, uLong p_filePathBufferSize = 0) -> int {
+		int err = unzGetCurrentFileInfo64(p_zip_file, &r_file_info, p_szFilePathBuffer, p_filePathBufferSize, nullptr, 0, nullptr, 0);
+		if (err != UNZ_OK) {
+			return err;
+		}
+
+		if (r_file_info.size_filename > p_filePathBufferSize) {
+			return UNZ_ERRNO; // filename doesn't fit in the buffer
+		}
+
+		// NOTE: p_szFilePathBuffer is not null-terminated when (file_info.size_filename == p_filePathBufferSize)
+		r_filepath = String::utf8(p_szFilePathBuffer, r_file_info.size_filename); // safely cast the file path
+
+		return UNZ_OK;
+	};
+
+	const uLong short_file_path_buffer_size = 16 * 1024ul;
```

Comment:
I had it originally at 1k but since 16k is what the rest of the engine uses I changed it to match that.

The `LocalVector` is only used in the "exceptional case", i.e. when it doesn't fit on the stack. This prevents repeat reading from the zip and missing the default fast/good path when most files are within a reasonable max path. It's only an encapsulation over the raw memory calls that are currently in use.

This is how things currently work without this PR. I've just tidied it up to, in-future, avoid repeating the mistakes that lead to the buffer-overread and leaking memory.

---

If anyone has any data on the path size constant or specifically wants to reduce it, I'm all for it.